### PR TITLE
added support for incomplete coords and incoming %-rgba values ...

### PR DESCRIPTION
**Pull request by ewerybody**
Heya! I found some issues that I needed to be fixed for our qss stuff to work after switching from the outdated [python-scss](https://github.com/pistolero/python-scss). Most notably:

- 8d7dd19 as Qt allows to fill in qlineargradient coordinates with 0 values automatically
- c6a89a5 %-values already in `rgba` statements are no longer multiplied by 100 again
- d6b827e now conform with former `compile` calls `compile_filename` also returns the resulting code.

cheers!
ëRiC

### CODE REVIEW

**Comment by ewerybody** [comment]
I'm sorry I don't get whats wrong here :/
It is probebly not the styling stuff, right? But where do I find what to do?

**Comment by ewerybody** [comment]
Ahhhh alrighty! 🌈

**Comment by danbradham** [comment]
Thanks for working on this @ewerybody, looks like it makes the qlineargradient conformer more robust.

https://github.com/spyder-ide/qtsass/commit/5b798edc1ae088df41df9c3ee470d1a38e317da7
Is there a reason that qtsass needs to be importable without watchdog? It's included as an install_requires in setup.py so it should be available by the time you import qtsass. I was under the impression that watchdog was available on all platforms. I think it makes sense to revert that commit for this PR, it doesn't seem to fit in with the primary issue this PR resolves.

@goanpeca, I'm not very familiar with the code formatting checks. Any ideas what's going on here with Travis?

**Comment by goanpeca** [comment]
Hi @ewerybody thanks for the work

>Is there a reason that qtsass needs to be importable without watchdog? It's included as an install_requires in setup.py so it should be available by the time you import qtsass. I was under the impression that watchdog was available on all platforms. I think it makes sense to revert that commit for this PR, it doesn't seem to fit in with the primary issue this PR resolves.

Agreed!

>@goanpeca, I'm not very familiar with the code formatting checks. Any ideas what's going on here with Travis?

I am bit swamped right now. I will look at this later today :-)

**Comment by ewerybody** [comment]
Heya! 

> Is there a reason that qtsass needs to be importable without watchdog? ... I think it makes sense to revert that commit for this PR, it doesn't seem to fit in with the primary issue this PR resolves.

Yeah true it's a mixed bag and this doesn't 'need' to be in here 👍 I'll try and revert that.
But in general I think this is a plus. `qtsass` itself is a nice compact addition to `libsass`.
Having to add `watchdog` also drags `pathtools` along and `argh` and `yaml` and whatnot ... If you create a portable package: I try to be reasonable about adding additional packages. And this is only for that watch functionality which we don't even need :)

**Comment by ewerybody** [comment]
> > @goanpeca, I'm not very familiar with the code formatting checks. Any ideas what's going on here with Travis?
> 
> I am bit swamped right now. I will look at this later today :-)

The latest travis report has this:
```
Please run the linter and formatter script!

 M qtsass/functions.py
```
But it does not say much more. Unfortunately I don't even see this locally on Python 2.7. 
Something removes the blank lines after class docstring and then pydocstyle D204 complains about it again 😐

**Comment by goanpeca** [comment]
>Yeah true it's a mixed bag and this doesn't 'need' to be in here 👍 I'll try and revert that.
But in general I think this is a plus. qtsass itself is a nice compact addition to libsass.
Having to add watchdog also drags pathtools along and argh and yaml and whatnot ... If you create a portable package: I try to be reasonable about adding additional packages. And this is only for that watch functionality which we don't even need :)

Agreed, we can make it optional then, make sure to update the `setup.py`

>The latest travis report has this:

I will fix this later today. Right now I am busy ;-p

**Comment by goanpeca** [comment]
@ewerybody could you add a test for both cases (ints and floats?) Thanks!

**Comment by ewerybody** [comment]
Oh I actually didn't want this to be part of the PR but thats automatically with anything that I push to my forks master right now, right? Sorry, I'm such a qithub n00b. That should be from a branch, right?

> @ewerybody could you add a test for both cases (ints and floats?) Thanks!

I added a test for the floats. Anything else is int already. Should that be in the same test right away?

**Comment by goanpeca** [comment]
>That should be from a branch, right?

Correct

But it should be easy to fix if you want to change it :-)

**Comment by goanpeca** [comment]
I think we are good to go. Let me do a final review before we do the merge (and yes for next time use a brach different from master ;-) )

File: run_checks_and_format.py [hunk 1]
```diff
@@ -30,7 +31,12 @@
 
 def run_process(cmd_list):
     """Run popen process."""
-    p = Popen(cmd_list, stdout=PIPE, stderr=PIPE)
+
+    try:
+        p = Popen(cmd_list, stdout=PIPE, stderr=PIPE)
+    except WindowsError:
```
Comment (goanpeca) [inline run_checks_and_format.py]: This is an specific windows error, so unless this only runs on windows this should be `OSError`

File: run_checks_and_format.py [hunk 2]
```diff
@@ -40,13 +46,22 @@ def run_process(cmd_list):
...
+    out, _err = run_process(['git', 'status', '--short'])
     out_lines = [l for l in out.split('\n') if l.strip()]
     return out_lines
 
 
 def run():
     """Run linters and formatters."""
+
+    # make sure scripts are available in environment paths
+    if sys.platform == 'win32':
```
Comment (goanpeca) [inline run_checks_and_format.py]: Why are we doing this here @ewerybody ?

File: tests/test_functions.py [hunk 3]
```diff
@@ -6,35 +6,41 @@
...
             'rgba(0, 1, 2, 30%)'
         )
 
+    def test_rgba_percentage_alpha(self):
+        result = self.compile_scss('rgba(255, 0, 125, 75%)')
+        self.assertEqual(result, 'rgba(255, 0, 125, 75%)')
+
+    def test_rgba_8bit_int_alpha(self):
+        result = self.compile_scss('rgba(255, 0, 125, 128)')
+        self.assertEqual(result, 'rgba(255, 0, 125, 50%)')
```
Comment (goanpeca) [inline tests/test_functions.py]: Let's test some "extremes" here (for 0 and 255)

File: run_checks_and_format.py (hunk 2 above)
Comment (ewerybody) [reply run_checks_and_format.py]: Well, I have multiple Python environments in parallel and the dev one with `pydocstyle`, `pycodestyle`, `yapf`, `isort` is not in the environment path. So these would not be available by just calling them right away.

Is there a better way to solve this? 🤔

Comment (goanpeca) [reply run_checks_and_format.py]: You will have to install either the linters on the same env as qtsass or viceversa. But this is a very specific hack for your setup that should not go in the package. Please remove it

Comment (ewerybody) [reply run_checks_and_format.py]: Hmm ... as you see in the code its just that the `Scripts` folder of the Python env I'm currently running is not in `PATH` env. So it does not matter where I have qtsass, does it?

... I would need a way to change the `Scripts` folder path in PATHs according to the Python env I wanna run... at best automatically. Maybe some scripts at first... 😔 well

**Comment by ewerybody** [comment]
btw. This is what `run_checks_and_format.py` spits out for me:
```
Running: pydocstyle qtsass
qtsass\conformers.py:25 in public class `Conformer`:
        D204: 1 blank line required after class docstring (found 0)
qtsass\conformers.py:36 in public class `NotConformer`:
        D204: 1 blank line required after class docstring (found 0)
qtsass\events.py:21 in public class `SourceEventHandler`:
        D204: 1 blank line required after class docstring (found 0)


Running: pycodestyle qtsass

Running: yapf qtsass --in-place --recursive

Running: isort -y
Skipped 1 files


All checks passed!
```

**Comment by goanpeca** [comment]
@ewerybody will merge as it is, thanks for working on this!. Please do not open more PRs from master. Make sure to create a branch always.

**Comment by ewerybody** [comment]
Happy to help! 😌

**Pull request merged**